File: sdk_typescript/src/origin.ts
import { refuse } from "./errors.ts";
import type { JsonValue } from "./json.ts";
import type { RecordType, VerificationKey } from "./types.ts";

export type PrimarySignerNamespace = VerificationKey["namespace"];

/** Closed ES-033 dispatch: a future record type has no signing default. */
export const PRIMARY_SIGNER_NAMESPACE = Object.freeze({
  AssuranceBoundary: "evidence",
  QualificationRecord: "evidence",
  PopulationRecord: "issuer",
  AgentIdentity: "evidence",
  ActionProposal: "evidence",
  AuthorityDecision: "evidence",
  HumanReview: "evidence",
  ExecutionReceipt: "evidence",
  ExternalConfirmation: "issuer",
  FinalityRecord: "evidence",
  OutcomeRecord: "evidence",
  CoverageGap: "evidence",
  AttestationWindow: "evidence",
  RevocationRecord: "issuer",
} satisfies Record<RecordType, PrimarySignerNamespace>);

export function primarySignerNamespace(recordType: JsonValue | undefined): PrimarySignerNamespace {
  if (typeof recordType !== "string" || !Object.hasOwn(PRIMARY_SIGNER_NAMESPACE, recordType)) {
    return refuse("schema.record_type_unknown", "record_type has no ES-033 origin classification");
  }
  return PRIMARY_SIGNER_NAMESPACE[recordType as RecordType];
}

File: sdk_python/evidence/origin.py
"""Closed record-origin signing policy (ES-033 / SE-003)."""

from __future__ import annotations

from collections.abc import Mapping
from types import MappingProxyType
from typing import Final, Literal

PrimarySignerNamespace = Literal["evidence", "issuer"]

# Complete rather than defaulting to evidence. A new record type is not
# authorised until the specification, this table, Go, and the vectors agree.
PRIMARY_SIGNER_NAMESPACE: Final[Mapping[str, PrimarySignerNamespace]] = MappingProxyType(
    {
        "AssuranceBoundary": "evidence",
        "QualificationRecord": "evidence",
        "PopulationRecord": "issuer",
        "AgentIdentity": "evidence",
        "ActionProposal": "evidence",
        "AuthorityDecision": "evidence",
        "HumanReview": "evidence",
        "ExecutionReceipt": "evidence",
        "ExternalConfirmation": "issuer",
        "FinalityRecord": "evidence",
        "OutcomeRecord": "evidence",
        "CoverageGap": "evidence",
        "AttestationWindow": "evidence",
        "RevocationRecord": "issuer",
    }
)


def primary_signer_namespace(record_type: str) -> PrimarySignerNamespace:
    """Return the role assigned by ES-033, with no permissive fallback."""

    try:
        return PRIMARY_SIGNER_NAMESPACE[record_type]
    except KeyError as exc:
        raise ValueError(
            f"record type {record_type!r} has no ES-033 origin classification"
        ) from exc
